### PR TITLE
Add edges for those unreachable nodes to improve graph connectivity

### PR DESCRIPTION
**Pull request by chasingegg**
issue #339 

GIST1M dataset, there are ~5000 unreachable nodes, we could search neighbors of these nodes to add reverse edges if there are space after we finish index building. After this pr, regular search does not change much

hnsw | Build time(s) | recall(top100) | qps
-- | -- | -- | --  
before this pr | 1200 | 0.911 | 2075    
after this pr | 1200(index building)+9(repair graph) | 0.9115 | 2036 

For these ~5000 unreachable nodes, we could never get themselves as the top1 results when we specify them as query vectors. After this pr, we could get a 0.99 recall of top1 under a small ef search param=30.

### CODE REVIEW

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 1]
```diff
@@ -1421,6 +1421,132 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+                tableint* datal = (tableint*)(data + 1);
+#if defined(USE_PREFETCH)
+                for (int i = 0; i < size; ++i) {
+                    _mm_prefetch(getDataByInternalId(datal[i]), _MM_HINT_T0);
+                }
+#endif
+                for (int i = 0; i < size; i++) {
+                    tableint cand = datal[i];
+                    if (cand < 0 || cand > max_elements_)
+                        throw std::runtime_error("cand error");
```
Comment (alexanderguzhva) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:1505]: how exactly is this exception handled in the parent code? A client gets 'Unknown error during building an index'?

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 2]
```diff
@@ -1421,6 +1421,132 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+                std::queue<tableint> q;
+                q.push(start_point);
+                used[start_point] = true;
+                if (level > 0)
+                    touched.push_back(start_point);
+                while (!q.empty()) {
+                    tableint j = q.front();
+                    q.pop();
+                    unsigned int* data;
+                    std::unique_lock<std::mutex> lock(link_list_locks_[j]);
```
[the remark this answers is not in the document]
Comment (chasingegg) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:1447]: Sure we can remove the lock.

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 3]
```diff
@@ -1421,6 +1421,132 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+        tableint currObj = enterpoint_node_;
+
+        dist_t curdist = calcDistance(cur_c, currObj);
+
+        for (int level_above = maxlevel_; level_above > level; level_above--) {
+            bool changed = true;
+            while (changed) {
+                changed = false;
+                unsigned int* data;
+                std::unique_lock<std::mutex> lock(link_list_locks_[currObj]);
```
[the remark this answers is not in the document]
Comment (chasingegg) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:1493]: Upper layer not changed, lock can be removed also

File: src/index/hnsw/hnsw.cc [hunk 4]
```diff
@@ -131,7 +131,21 @@ class HnswIndexNode : public IndexNode {
             futures.clear();
         }
 
-        build_time.RecordSection("");
+        build_time.RecordSection("graph build");
+        std::vector<unsigned> unreached = index_->findUnreachableVectors();
```
[the remark this answers is not in the document]
Comment (chasingegg) [reply src/index/hnsw/hnsw.cc:135]: It is always needed, to check if there are unreachable nodes after index building

File: thirdparty/hnswlib/hnswlib/hnswalg.h (hunk 1 above)
Comment (chasingegg) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:1505]: This code is duplicate with hnswlib search upper layer related code, from my understanding, this should never happen. However, there are bunch of runtime error in the hnswalg.h and caller never handle these situations. I will add exception handler in the Add() function

File: src/index/hnsw/hnsw.cc [hunk 5]
```diff
@@ -131,7 +131,21 @@ class HnswIndexNode : public IndexNode {
...
+        int unreached_num = unreached.size();
+        LOG_KNOWHERE_INFO_ << "there are " << unreached_num << " points can not be reached";
+        if (unreached_num > 0) {
+            futures.reserve(unreached_num);
+            for (int i = 0; i < unreached_num; ++i) {
```
[the remark this answers is not in the document]
Comment (chasingegg) [reply src/index/hnsw/hnsw.cc:140]: For the GIST1M dataset, ~5000 unreachable nodes, If repair it in parallel we can get about repair time from 20s->9s. Although these unreachable nodes are mostly not too many, I suggest to parallel it to cover these corner cases, and we can reduce repair time in the common cases also.

[the remark this answers is not in the document]
Comment (alexanderguzhva) [reply src/index/hnsw/hnsw.cc]: Does it make sense to add a threshold, something like `if (unreached_num > threshold)` ?

File: src/index/hnsw/hnsw.cc [hunk 6]
```diff
@@ -109,32 +109,52 @@ class HnswIndexNode : public IndexNode {
...
+        try {
+            index_->addPoint(tensor, 0);
+
+            futures.reserve(batch_size);
+            for (int64_t round_id = 0; round_id < round_num; round_id++) {
+                int64_t start_id = (shuffle_build ? shuffle_batch_ids[round_id] : round_id) * batch_size;
+                int64_t end_id =
+                    std::min(rows - 1, ((shuffle_build ? shuffle_batch_ids[round_id] : round_id) + 1) * batch_size);
+                for (int64_t i = start_id; i < end_id; ++i) {
+                    futures.emplace_back(build_pool->push([&, idx = i + 1]() {
```
Comment (alexanderguzhva) [inline src/index/hnsw/hnsw.cc:121]: hmm, the typical use case is somewhat inverse:
```
for (int64_t round_id = 0; round_id < round_num; round_id++) {
    futures.emplace_back(build_pool->push([&, local_round_id=round_id]() {
        // process from start_id till end_id
        for (int64_t i = start_id; i < end_id; i++) {...}
    }));
}
for (auto& future : futures)
    future.wait();
}

```

File: src/index/hnsw/hnsw.cc [hunk 7]
```diff
@@ -109,32 +109,52 @@ class HnswIndexNode : public IndexNode {
...
+                        uint64_t added = counter.fetch_add(1);
+                        if (added % one_tenth_row == 0) {
+                            LOG_KNOWHERE_INFO_ << "HNSW build progress: " << (added / one_tenth_row) << "0%";
+                        }
+                    }));
+                }
+                for (auto& future : futures) {
+                    future.wait();
+                }
+                futures.clear();
```
Comment (alexanderguzhva) [inline src/index/hnsw/hnsw.cc:132]: Please add a check for exceptions:
```
    // wait until completion
    for (auto& future : futures) {
        future.wait();
    }

    // check for exceptions
    for (auto& future : futures) {
        future.result().value();
    }
```

File: src/index/hnsw/hnsw.cc [hunk 8]
```diff
@@ -109,32 +109,52 @@ class HnswIndexNode : public IndexNode {
...
+            std::vector<unsigned> unreached = index_->findUnreachableVectors();
+            int unreached_num = unreached.size();
+            LOG_KNOWHERE_INFO_ << "there are " << unreached_num << " points can not be reached";
+            if (unreached_num > 0) {
+                futures.reserve(unreached_num);
+                for (int i = 0; i < unreached_num; ++i) {
+                    futures.emplace_back(
+                        build_pool->push([&, idx = i]() { index_->repairGraphConnectivity(unreached[idx]); }));
+                }
+                for (auto& future : futures) {
```
Comment (alexanderguzhva) [inline src/index/hnsw/hnsw.cc:145]:
```
    // wait until completion
    for (auto& future : futures) {
        future.wait();
    }

    // check for exceptions
    for (auto& future : futures) {
        future.result().value();
    }
```

**Pull request merged**